File: server/evaluation/python/campaign_evaluator.py
#!/usr/bin/env python3
"""
A3: Campaign Evaluator

Evaluates campaign performance based on ROAS7 and RET7 achievement rates.
Generates recommendations and action options for optimizers.
"""

import sys
from datetime import datetime
from typing import Dict, Any, List, Optional
from db_utils import get_db, format_output, read_input


class CampaignEvaluator:
    """Evaluate campaign performance and generate recommendations"""

    # Campaign type thresholds
    TEST_CAMPAIGN_THRESHOLD = 1000.0  # Total spend < $1000 = test campaign

    # Achievement rate thresholds
    DANGER_THRESHOLD = 60.0  # < 60% = danger
    WARNING_THRESHOLD = 85.0  # 60-85% = warning
    OBSERVATION_THRESHOLD = 100.0  # 85-100% = observation
    HEALTHY_THRESHOLD = 110.0  # 100-110% = healthy, >= 110% = excellent

    def __init__(self):
        self.db = get_db()

    def evaluate_campaign(
        self,
        campaign_id: str,
        evaluation_date: Optional[str] = None
    ) -> Dict[str, Any]:
        """
        Evaluate a single campaign

        Args:
            campaign_id: Campaign ID
            evaluation_date: Evaluation date in ISO format (default: today)

        Returns:
            Dictionary containing:
            - campaign_id: str
            - campaign_name: str
            - campaign_type: "test" | "mature"
            - total_spend: float
            - actual_roas7: float
            - actual_ret7: float
            - baseline_roas7: float
            - baseline_ret7: float
            - roas_achievement_rate: float (%)
            - ret_achievement_rate: float (%)
            - min_achievement_rate: float (%)
            - recommendation_type: str
            - status: "danger" | "warning" | "observation" | "healthy" | "excellent"
            - action_options: List of available actions
        """
        try:
            # Parse evaluation date
            if evaluation_date:
                eval_date = datetime.fromisoformat(evaluation_date.replace('Z', '+00:00'))
            else:
                eval_date = datetime.now()

            eval_date_str = eval_date.strftime("%Y-%m-%d")

            with self.db:
                # 1. Get campaign basic info and performance data
                campaign_query = """
                    SELECT
                        campaign_id,
                        campaign_name,
                        product_name,
                        country_code,
                        platform,
                        channel,
                        total_spend,
                        actual_roas7,
                        actual_ret7
                    FROM mock_campaign_performance
                    WHERE campaign_id = %s
                      AND date = %s
                    LIMIT 1
                """

                campaign_results = self.db.execute_query(campaign_query, (campaign_id, eval_date_str))

                if not campaign_results or len(campaign_results) == 0:
                    return {
                        "error": f"Campaign {campaign_id} not found for date {eval_date_str}",
                        "campaign_id": campaign_id
                    }

                campaign = campaign_results[0]

                # 2. Determine campaign type
                total_spend = float(campaign['total_spend'])
                campaign_type = "test" if total_spend < self.TEST_CAMPAIGN_THRESHOLD else "mature"

                # 3. Get safety baseline
                baseline_query = """
                    SELECT baseline_roas7, baseline_ret7, reference_period
                    FROM safety_baseline
                    WHERE product_name = %s
                      AND country_code = %s
                      AND platform = %s
                      AND channel = %s
                    LIMIT 1
                """

                baseline_results = self.db.execute_query(
                    baseline_query,
                    (campaign['product_name'], campaign['country_code'],
                     campaign['platform'], campaign['channel'])
                )

                if not baseline_results or len(baseline_results) == 0:
                    return {
                        "error": f"No baseline found for {campaign['product_name']}/{campaign['country_code']}",
                        "campaign_id": campaign_id
                    }

                baseline = baseline_results[0]
                baseline_roas7 = float(baseline['baseline_roas7'])
                baseline_ret7 = float(baseline['baseline_ret7'])

                # 4. Calculate achievement rates
                actual_roas7 = float(campaign['actual_roas7'])
                actual_ret7 = float(campaign['actual_ret7'])

                roas_achievement_rate = (actual_roas7 / baseline_roas7 * 100) if baseline_roas7 > 0 else 0.0
                ret_achievement_rate = (actual_ret7 / baseline_ret7 * 100) if baseline_ret7 > 0 else 0.0

                # Use minimum achievement rate (bucket effect - weakest link determines overall health)
                min_achievement_rate = min(roas_achievement_rate, ret_achievement_rate)

                # 5. Generate recommendation
                recommendation = self.generate_recommendation(min_achievement_rate)
                status = self.get_status(min_achievement_rate)

                # 6. Generate action options
                action_options = self.generate_action_options(recommendation['type'])

                # 7. Save evaluation to database
                self.save_evaluation(
                    campaign_id=campaign_id,
                    campaign_name=campaign['campaign_name'],
                    evaluation_date=eval_date_str,
                    campaign_type=campaign_type,
                    total_spend=total_spend,
                    actual_roas7=actual_roas7,
                    actual_ret7=actual_ret7,
                    baseline_roas7=baseline_roas7,
                    baseline_ret7=baseline_ret7,
                    roas_achievement_rate=roas_achievement_rate,
                    ret_achievement_rate=ret_achievement_rate,
                    min_achievement_rate=min_achievement_rate,
                    recommendation_type=recommendation['type'],
                    status=status
                )

                return {
                    "campaign_id": campaign_id,
                    "campaign_name": campaign['campaign_name'],
                    "campaign_type": campaign_type,
                    "total_spend": total_spend,
                    "actual_roas7": actual_roas7,
                    "actual_ret7": actual_ret7,
                    "baseline_roas7": baseline_roas7,
                    "baseline_ret7": baseline_ret7,
                    "roas_achievement_rate": round(roas_achievement_rate, 2),
                    "ret_achievement_rate": round(ret_achievement_rate, 2),
                    "min_achievement_rate": round(min_achievement_rate, 2),
                    "recommendation_type": recommendation['type'],
                    "recommendation_desc": recommendation['description'],
                    "status": status,
                    "action_options": action_options
                }

        except Exception as e:
            print(f"Campaign evaluation error: {e}", file=sys.stderr, flush=True)
            return {
                "error": str(e),
                "campaign_id": campaign_id
            }

    def generate_recommendation(self, min_achievement_rate: float) -> Dict[str, str]:
        """
        Generate recommendation based on achievement rate

        Args:
            min_achievement_rate: Minimum of ROAS and RET achievement rates

        Returns:
            Dictionary with type and description
        """
        if min_achievement_rate < self.DANGER_THRESHOLD:
            return {
                "type": "关停",
                "description": "达成率低于60%，严重亏损，建议立即关停campaign"
            }
        elif min_achievement_rate < self.WARNING_THRESHOLD:
            return {
                "type": "保守缩量",
                "description": "达成率60-85%，表现不佳，建议减少投入观察变化"
            }
        elif min_achievement_rate < self.OBSERVATION_THRESHOLD:
            return {
                "type": "继续观察",
                "description": "达成率85-100%，接近及格线，保持现状等待更多数据"
            }
        elif min_achievement_rate < self.HEALTHY_THRESHOLD:
            return {
                "type": "保守扩量或观察",
                "description": "达成率100-110%，表现达标，可小幅增加投入测试"
            }
        else:
            return {
                "type": "激进扩量",
                "description": "达成率≥110%，表现优异，建议大幅增加投入"
            }

    def get_status(self, min_achievement_rate: float) -> str:
        """Get status label based on achievement rate"""
        if min_achievement_rate < self.DANGER_THRESHOLD:
            return "danger"
        elif min_achievement_rate < self.WARNING_THRESHOLD:
            return "warning"
        elif min_achievement_rate < self.OBSERVATION_THRESHOLD:
            return "observation"
        elif min_achievement_rate < self.HEALTHY_THRESHOLD:
            return "healthy"
        else:
            return "excellent"

    def generate_action_options(self, recommendation_type: str) -> List[Dict[str, Any]]:
        """
        Generate action options based on recommendation type

        Args:
            recommendation_type: Type of recommendation

        Returns:
            List of action option groups
        """
        if recommendation_type in ["激进扩量", "保守扩量或观察"]:
            # Scale up actions
            return [
                {
                    "type": "budget",
                    "label": "提高预算",
                    "options": [
                        {"value": "+1%", "description": "试探性调整"},
                        {"value": "+3%", "description": "标准调整"},
                        {"value": "+5%", "description": "激进调整"},
                        {"value": "custom", "description": "自定义"}
                    ]
                },
                {
                    "type": "troas",
                    "label": "降低tROAS",
                    "options": [
                        {"value": "-1%", "description": "试探性调整"},
                        {"value": "-3%", "description": "标准调整"},
                        {"value": "-5%", "description": "激进调整"},
                        {"value": "custom", "description": "自定义"}
                    ]
                },
                {
                    "type": "none",
                    "label": "暂不调整",
                    "options": [{"value": "observe", "description": "继续观察"}]
                }
            ]
        elif recommendation_type in ["保守缩量", "继续观察"]:
            # Scale down or observe actions
            return [
                {
                    "type": "budget",
                    "label": "降低预算",
                    "options": [
                        {"value": "-1%", "description": "试探性调整"},
                        {"value": "-3%", "description": "标准调整"},
                        {"value": "-5%", "description": "激进调整"},
                        {"value": "custom", "description": "自定义"}
                    ]
                },
                {
                    "type": "troas",
                    "label": "提高tROAS",
                    "options": [
                        {"value": "+1%", "description": "试探性调整"},
                        {"value": "+3%", "description": "标准调整"},
                        {"value": "+5%", "description": "激进调整"},
                        {"value": "custom", "description": "自定义"}
                    ]
                },
                {
                    "type": "pause",
                    "label": "暂停campaign",
                    "options": [{"value": "pause", "description": "立即暂停"}]
                },
                {
                    "type": "none",
                    "label": "继续观察",
                    "options": [{"value": "observe", "description": "暂不调整"}]
                }
            ]
        else:  # 关停
            return [
                {
                    "type": "pause",
                    "label": "暂停campaign",
                    "options": [{"value": "pause", "description": "立即暂停"}]
                },
                {
                    "type": "none",
                    "label": "继续观察",
                    "options": [{"value": "observe", "description": "暂不调整，再观察"}]
                }
            ]

    def save_evaluation(
        self,
        campaign_id: str,
        campaign_name: str,
        evaluation_date: str,
        campaign_type: str,
        total_spend: float,
        actual_roas7: float,
        actual_ret7: float,
        baseline_roas7: float,
        baseline_ret7: float,
        roas_achievement_rate: float,
        ret_achievement_rate: float,
        min_achievement_rate: float,
        recommendation_type: str,
        status: str
    ) -> bool:
        """Save evaluation result to database"""
        try:
            query = """
                INSERT INTO campaign_evaluation (
                    campaign_id, campaign_name, evaluation_date, campaign_type,
                    total_spend, actual_roas7, actual_ret7,
                    baseline_roas7, baseline_ret7,
                    roas_achievement_rate, ret_achievement_rate, min_achievement_rate,
                    recommendation_type, status, created_at
                )
                VALUES (%s, %s, %s, %s, %s, %s, %s, %s, %s, %s, %s, %s, %s, %s, NOW())
            """

            return self.db.execute_update(
                query,
                (campaign_id, campaign_name, evaluation_date, campaign_type,
                 total_spend, actual_roas7, actual_ret7,
                 baseline_roas7, baseline_ret7,
                 roas_achievement_rate, ret_achievement_rate, min_achievement_rate,
                 recommendation_type, status)
            )

        except Exception as e:
            print(f"Save evaluation error: {e}", file=sys.stderr, flush=True)
            return False

    def evaluate_all_campaigns(self, evaluation_date: Optional[str] = None) -> Dict[str, Any]:
        """
        Batch evaluate all campaigns

        Args:
            evaluation_date: Evaluation date (default: today)

        Returns:
            Summary of evaluation results
        """
        try:
            # Parse evaluation date
            if evaluation_date:
                eval_date = datetime.fromisoformat(evaluation_date.replace('Z', '+00:00'))
            else:
                eval_date = datetime.now()

            eval_date_str = eval_date.strftime("%Y-%m-%d")

            with self.db:
                # Get all campaigns for the evaluation date
                campaigns_query = """
                    SELECT DISTINCT campaign_id
                    FROM mock_campaign_performance
                    WHERE date = %s
                    ORDER BY campaign_id
                """

                campaigns = self.db.execute_query(campaigns_query, (eval_date_str,))

                results = []
                success_count = 0
                failed_count = 0

                for campaign in campaigns:
                    campaign_id = campaign['campaign_id']
                    evaluation = self.evaluate_campaign(campaign_id, evaluation_date)

                    if 'error' not in evaluation:
                        success_count += 1
                        results.append({
                            "campaign_id": campaign_id,
                            "status": evaluation['status'],
                            "min_achievement_rate": evaluation['min_achievement_rate'],
                            "recommendation": evaluation['recommendation_type']
                        })
                    else:
                        failed_count += 1
                        results.append({
                            "campaign_id": campaign_id,
                            "error": evaluation['error']
                        })

                # Group by status
                status_summary = {
                    "danger": len([r for r in results if r.get('status') == 'danger']),
                    "warning": len([r for r in results if r.get('status') == 'warning']),
                    "observation": len([r for r in results if r.get('status') == 'observation']),
                    "healthy": len([r for r in results if r.get('status') == 'healthy']),
                    "excellent": len([r for r in results if r.get('status') == 'excellent'])
                }

                return {
                    "success": True,
                    "evaluation_date": eval_date_str,
                    "total_campaigns": len(campaigns),
                    "success_count": success_count,
                    "failed_count": failed_count,
                    "status_summary": status_summary,
                    "results": results
                }

        except Exception as e:
            print(f"Batch evaluation error: {e}", file=sys.stderr, flush=True)
            return {
                "success": False,
                "error": str(e)
            }


def main():
    """Main entry point for CLI usage"""
    input_data = read_input()

    evaluator = CampaignEvaluator()

    action = input_data.get('action', 'evaluate')

    if action == 'evaluate':
        # Evaluate single campaign
        result = evaluator.evaluate_campaign(
            campaign_id=input_data.get('campaignId'),
            evaluation_date=input_data.get('evaluationDate')
        )
    elif action == 'evaluate_all':
        # Batch evaluate all campaigns
        result = evaluator.evaluate_all_campaigns(
            evaluation_date=input_data.get('evaluationDate')
        )
    else:
        result = {"error": f"Unknown action: {action}"}

    format_output(result)


if __name__ == "__main__":
    main()

File: server/evaluation/wrappers/campaign-evaluator.ts
/**
 * A3: Campaign Evaluator - TypeScript Wrapper
 *
 * Phase 5: Refactored to use real AppsFlyer cohort data.
 * The Python-based evaluation is kept for backward compatibility but deprecated.
 */

import { spawn } from 'child_process'
import path from 'path'
import {
  getAggregatedCampaignMetrics,
  getCampaignsFromAF,
} from '../../db/queries-evaluation'
import { getSafetyBaseline, createCampaignEvaluation } from '../../db/queries-evaluation'
import { calculateBaselineFromAF, getOrCreateBaselineSettings } from './baseline-calculator'

// ============================================
// TYPE DEFINITIONS
// ============================================

export type CampaignStatus = 'danger' | 'warning' | 'observation' | 'healthy' | 'excellent'

export interface CampaignEvaluationResult {
  campaign_id: string
  campaign_name: string
  campaign_type: 'test' | 'mature'
  total_spend: number
  actual_roas7: number | null
  actual_ret7: number | null
  baseline_roas7: number | null
  baseline_ret7: number | null
  roas_achievement_rate: number | null
  ret_achievement_rate: number | null
  min_achievement_rate: number | null
  recommendation_type: string
  recommendation_desc: string
  status: CampaignStatus
  action_options: ActionOption[]
  error?: string
  dataSource?: 'appsflyer' | 'mock'
  hasData?: boolean
}

export interface ActionOption {
  type: string
  label: string
  options: Array<{
    value: string
    description: string
  }>
}

export interface BatchEvaluationResult {
  success: boolean
  evaluation_date: string
  total_campaigns: number
  success_count: number
  failed_count: number
  status_summary: {
    danger: number
    warning: number
    observation: number
    healthy: number
    excellent: number
  }
  results: Array<{
    campaign_id: string
    status?: string
    min_achievement_rate?: number
    recommendation?: string
    error?: string
  }>
  error?: string
  dataSource?: 'appsflyer' | 'mock'
}

export interface EvaluateCampaignFromAFParams {
  campaign: string
  appId: string
  geo: string
  mediaSource: string
  evaluationDate?: Date
}

// ============================================
// HELPER FUNCTIONS
// ============================================

/**
 * Map achievement rate to campaign status
 * Per PRD thresholds
 */
function mapAchievementToStatus(rate: number | null): CampaignStatus {
  if (rate === null) return 'observation'
  if (rate >= 110) return 'excellent'
  if (rate >= 100) return 'healthy'
  if (rate >= 85) return 'observation'
  if (rate >= 60) return 'warning'
  return 'danger'
}

/**
 * Map status to recommendation
 */
function mapStatusToRecommendation(status: CampaignStatus): {
  type: string
  desc: string
  options: ActionOption[]
} {
  switch (status) {
    case 'excellent':
      return {
        type: 'scale_up',
        desc: 'Campaign performing exceptionally well. Consider scaling budget.',
        options: [
          {
            type: 'budget',
            label: 'Budget Adjustment',
            options: [
              { value: 'increase_20', description: 'Increase budget by 20%' },
              { value: 'increase_50', description: 'Increase budget by 50%' },
              { value: 'maintain', description: 'Maintain current budget' },
            ],
          },
        ],
      }
    case 'healthy':
      return {
        type: 'maintain',
        desc: 'Campaign performing at baseline. Monitor and maintain.',
        options: [
          {
            type: 'monitoring',
            label: 'Monitoring Strategy',
            options: [
              { value: 'weekly_review', description: 'Weekly performance review' },
              { value: 'maintain', description: 'Maintain current strategy' },
            ],
          },
        ],
      }
    case 'observation':
      return {
        type: 'observe',
        desc: 'Campaign slightly below baseline. Watch closely.',
        options: [
          {
            type: 'adjustment',
            label: 'Adjustment Options',
            options: [
              { value: 'optimize_creative', description: 'Optimize creatives' },
              { value: 'adjust_targeting', description: 'Adjust targeting' },
              { value: 'wait', description: 'Wait for more data' },
            ],
          },
        ],
      }
    case 'warning':
      return {
        type: 'optimize',
        desc: 'Campaign underperforming. Optimization required.',
        options: [
          {
            type: 'action',
            label: 'Recommended Actions',
            options: [
              { value: 'reduce_budget', description: 'Reduce budget by 30%' },
              { value: 'pause_creatives', description: 'Pause poor creatives' },
              { value: 'revise_strategy', description: 'Revise campaign strategy' },
            ],
          },
        ],
      }
    case 'danger':
      return {
        type: 'urgent_action',
        desc: 'Campaign critically underperforming. Immediate action required.',
        options: [
          {
            type: 'action',
            label: 'Urgent Actions',
            options: [
              { value: 'pause_campaign', description: 'Pause campaign' },
              { value: 'reduce_budget_50', description: 'Reduce budget by 50%' },
              { value: 'complete_review', description: 'Complete strategy review' },
            ],
          },
        ],
      }
  }
}

// ============================================
// APPSFLYER-BASED EVALUATION (PRIMARY)
// ============================================

/**
 * Evaluate a campaign using real AppsFlyer cohort data
 *
 * This is the primary function for Phase 5+, replacing mock data.
 *
 * @param params - Campaign evaluation parameters
 * @returns Campaign evaluation result with recommendations
 *
 * @example
 * ```typescript
 * const result = await evaluateCampaignFromAF({
 *   campaign: "my_campaign_name",
 *   appId: "id123456789",
 *   geo: "US",
 *   mediaSource: "googleadwords_int",
 * });
 * if (result.status === "danger") {
 *   console.log("Campaign needs attention!");
 * }
 * ```
 */
export async function evaluateCampaignFromAF(
  params: EvaluateCampaignFromAFParams
): Promise<CampaignEvaluationResult> {
  const { campaign, appId, geo, mediaSource, evaluationDate = new Date() } = params

  try {
    // Get cohort metrics for recent install dates (last 7 days of installs)
    const endDate = evaluationDate
    const startDate = new Date(evaluationDate)
    startDate.setDate(startDate.getDate() - 7)

    const metrics = await getAggregatedCampaignMetrics({
      campaign,
      appId,
      geo,
      mediaSource,
      installDateStart: startDate,
      installDateEnd: endDate,
    })

    // Get or calculate baseline
    let baseline = await getSafetyBaseline({
      productName: appId,
      countryCode: geo,
      platform: 'AppsFlyer',
      channel: mediaSource,
    })

    if (!baseline) {
      // Calculate baseline if not exists
      const baselineResult = await calculateBaselineFromAF({ appId, geo, mediaSource })
      if (baselineResult.hasData) {
        baseline = {
          id: 0,
          productName: appId,
          countryCode: geo,
          platform: 'AppsFlyer',
          channel: mediaSource,
          baselineRoas7: baselineResult.baseline_roas7?.toString() || null,
          baselineRet7: baselineResult.baseline_ret7?.toString() || null,
          referencePeriod: baselineResult.reference_period,
          lastUpdated: new Date(),
        }
      }
    }

    const baselineRoas7 = baseline?.baselineRoas7 ? parseFloat(baseline.baselineRoas7) : null
    const baselineRet7 = baseline?.baselineRet7 ? parseFloat(baseline.baselineRet7) : null

    // Calculate achievement rates
    let roasAchievementRate: number | null = null
    let retAchievementRate: number | null = null

    if (metrics.roas !== null && baselineRoas7 && baselineRoas7 > 0) {
      roasAchievementRate = (metrics.roas / baselineRoas7) * 100
    }

    if (metrics.retention !== null && baselineRet7 && baselineRet7 > 0) {
      retAchievementRate = (metrics.retention / baselineRet7) * 100
    }

    // Min achievement rate
    let minAchievementRate: number | null = null
    if (roasAchievementRate !== null && retAchievementRate !== null) {
      minAchievementRate = Math.min(roasAchievementRate, retAchievementRate)
    } else if (roasAchievementRate !== null) {
      minAchievementRate = roasAchievementRate
    } else if (retAchievementRate !== null) {
      minAchievementRate = retAchievementRate
    }

    // Determine status and recommendation
    const status = mapAchievementToStatus(minAchievementRate)
    const recommendation = mapStatusToRecommendation(status)

    // Determine campaign type (mature if > 30 days of data)
    const campaignType = metrics.cohortCount >= 30 ? 'mature' : 'test'

    // Store evaluation result
    if (minAchievementRate !== null) {
      await createCampaignEvaluation({
        campaignId: campaign,
        evaluationDate: evaluationDate.toISOString().split('T')[0],
        actualRoas7: metrics.roas?.toString() || null,
        actualRet7: metrics.retention?.toString() || null,
        roasAchievementRate: roasAchievementRate?.toString() || null,
        retAchievementRate: retAchievementRate?.toString() || null,
        minAchievementRate: minAchievementRate?.toString() || null,
        status,
        recommendationType: recommendation.type,
      })
    }

    return {
      campaign_id: campaign,
      campaign_name: campaign,
      campaign_type: campaignType,
      total_spend: metrics.totalCostUsd,
      actual_roas7: metrics.roas,
      actual_ret7: metrics.retention,
      baseline_roas7: baselineRoas7,
      baseline_ret7: baselineRet7,
      roas_achievement_rate: roasAchievementRate,
      ret_achievement_rate: retAchievementRate,
      min_achievement_rate: minAchievementRate,
      recommendation_type: recommendation.type,
      recommendation_desc: recommendation.desc,
      status,
      action_options: recommendation.options,
      dataSource: 'appsflyer',
      hasData: metrics.cohortCount > 0,
    }
  } catch (error) {
    const errorMessage = error instanceof Error ? error.message : 'Unknown error'
    return {
      campaign_id: campaign,
      campaign_name: campaign,
      campaign_type: 'test',
      total_spend: 0,
      actual_roas7: null,
      actual_ret7: null,
      baseline_roas7: null,
      baseline_ret7: null,
      roas_achievement_rate: null,
      ret_achievement_rate: null,
      min_achievement_rate: null,
      recommendation_type: 'error',
      recommendation_desc: `Evaluation failed: ${errorMessage}`,
      status: 'observation',
      action_options: [],
      error: errorMessage,
      dataSource: 'appsflyer',
      hasData: false,
    }
  }
}

/**
 * Batch evaluate all campaigns from AppsFlyer data
 *
 * @param params - Evaluation parameters
 * @returns Batch evaluation summary with status breakdown
 */
export async function evaluateAllCampaignsFromAF(params?: {
  appId?: string
  geo?: string
  mediaSource?: string
  evaluationDate?: Date
}): Promise<BatchEvaluationResult> {
  const { appId, geo, mediaSource, evaluationDate = new Date() } = params || {}

  try {
    // Get date range for campaign discovery (last 30 days)
    const endDate = evaluationDate
    const startDate = new Date(evaluationDate)
    startDate.setDate(startDate.getDate() - 30)

    // Get all campaigns from AppsFlyer
    const campaigns = await getCampaignsFromAF({
      appId,
      geo,
      mediaSource,
      installDateStart: startDate,
      installDateEnd: endDate,
    })

    const results: BatchEvaluationResult['results'] = []
    const statusSummary = {
      danger: 0,
      warning: 0,
      observation: 0,
      healthy: 0,
      excellent: 0,
    }
    let successCount = 0
    let failedCount = 0

    // Evaluate each campaign
    for (const campaign of campaigns) {
      try {
        const evaluation = await evaluateCampaignFromAF({
          campaign: campaign.campaign,
          appId: campaign.appId,
          geo: campaign.geo,
          mediaSource: campaign.mediaSource,
          evaluationDate,
        })

        if (evaluation.hasData) {
          statusSummary[evaluation.status]++
          results.push({
            campaign_id: campaign.campaign,
            status: evaluation.status,
            min_achievement_rate: evaluation.min_achievement_rate ?? undefined,
            recommendation: evaluation.recommendation_type,
          })
          successCount++
        } else {
          results.push({
            campaign_id: campaign.campaign,
            error: 'Insufficient data',
          })
          failedCount++
        }
      } catch (error) {
        const errorMessage = error instanceof Error ? error.message : 'Unknown error'
        results.push({
          campaign_id: campaign.campaign,
          error: errorMessage,
        })
        failedCount++
      }
    }

    return {
      success: failedCount === 0,
      evaluation_date: evaluationDate.toISOString().split('T')[0],
      total_campaigns: campaigns.length,
      success_count: successCount,
      failed_count: failedCount,
      status_summary: statusSummary,
      results,
      dataSource: 'appsflyer',
    }
  } catch (error) {
    const errorMessage = error instanceof Error ? error.message : 'Unknown error'
    return {
      success: false,
      evaluation_date: evaluationDate.toISOString().split('T')[0],
      total_campaigns: 0,
      success_count: 0,
      failed_count: 0,
      status_summary: {
        danger: 0,
        warning: 0,
        observation: 0,
        healthy: 0,
        excellent: 0,
      },
      results: [],
      error: `Batch evaluation failed: ${errorMessage}`,
      dataSource: 'appsflyer',
    }
  }
}

// ============================================
// LEGACY PYTHON-BASED FUNCTIONS (DEPRECATED)
// ============================================

/**
 * Run Python script and return parsed JSON output
 *
 * @deprecated Use evaluateCampaignFromAF instead
 */
async function runPythonScript<T>(scriptName: string, input: Record<string, unknown>): Promise<T> {
  console.warn(
    '[DEPRECATED] Using Python-based campaign evaluation. Switch to evaluateCampaignFromAF for real AppsFlyer data.'
  )

  return new Promise((resolve, reject) => {
    const scriptPath = path.join(process.cwd(), 'server', 'evaluation', 'python', scriptName)

    const pythonProcess = spawn('python3', [scriptPath])

    let stdoutData = ''
    let stderrData = ''

    pythonProcess.stdout.on('data', (data) => {
      stdoutData += data.toString()
    })

    pythonProcess.stderr.on('data', (data) => {
      stderrData += data.toString()
    })

    pythonProcess.on('close', (code) => {
      if (code !== 0) {
        reject(new Error(`Python script exited with code ${code}\nStderr: ${stderrData}`))
        return
      }

      try {
        const result = JSON.parse(stdoutData)
        resolve(result)
      } catch (error) {
        reject(
          new Error(
            `Failed to parse Python output: ${error}\nOutput: ${stdoutData}\nStderr: ${stderrData}`
          )
        )
      }
    })

    pythonProcess.on('error', (error) => {
      reject(new Error(`Failed to start Python process: ${error.message}`))
    })

    pythonProcess.stdin.write(JSON.stringify(input))
    pythonProcess.stdin.end()
  })
}

/**
 * Evaluate a single campaign
 *
 * @deprecated Use evaluateCampaignFromAF instead. This function uses mock data via Python.
 *
 * Calculates ROAS7/RET7 achievement rates and generates recommendations
 *
 * @param campaignId - Campaign ID to evaluate
 * @param evaluationDate - Date to evaluate (default: today)
 * @returns Campaign evaluation result with recommendations
 */
export async function evaluateCampaign(
  campaignId: string,
  evaluationDate?: Date | string
): Promise<CampaignEvaluationResult> {
  console.warn('[DEPRECATED] evaluateCampaign uses mock data. Use evaluateCampaignFromAF instead.')

  // Handle both Date objects and date strings
  let dateStr: string | undefined
  if (evaluationDate) {
    if (typeof evaluationDate === 'string') {
      dateStr = evaluationDate
    } else {
      dateStr = evaluationDate.toISOString()
    }
  }

  const input = {
    action: 'evaluate',
    campaignId,
    evaluationDate: dateStr,
  }

  const result = await runPythonScript<CampaignEvaluationResult>('campaign_evaluator.py', input)
  return { ...result, dataSource: 'mock' }
}

/**
 * Batch evaluate all campaigns
 *
 * @deprecated Use evaluateAllCampaignsFromAF instead. This function uses mock data via Python.
 *
 * Evaluates all campaigns for a given date and returns summary statistics
 *
 * @param evaluationDate - Date to evaluate (default: today)
 * @returns Batch evaluation summary with status breakdown
 */
export async function evaluateAllCampaigns(evaluationDate?: Date): Promise<BatchEvaluationResult> {
  console.warn(
    '[DEPRECATED] evaluateAllCampaigns uses mock data. Use evaluateAllCampaignsFromAF instead.'
  )

  const input = {
    action: 'evaluate_all',
    evaluationDate: evaluationDate?.toISOString(),
  }

  const result = await runPythonScript<BatchEvaluationResult>('campaign_evaluator.py', input)
  return { ...result, dataSource: 'mock' }
}

/**
 * Re-export database functions for direct access
 */
export { getCampaignEvaluations, createCampaignEvaluation } from '../../db/queries-evaluation'
